src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -3139,10 +3139,24 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           val initElems = scope.elems
           // SI-5877 The decls of a package include decls of the package object. But we don't want to add
           //         the corresponding synthetics to the package class, only to the package object class.
-          def shouldAdd(sym: Symbol) =
-            inBlock || !context.isInPackageObject(sym, context.owner)
+          // SI-6734 Locality test below is meaningless if we're not even in the correct tree.
+          def shouldAdd(sym: Symbol): Boolean = {
+            def shouldAddAsModule: Boolean =
+              sym.moduleClass.attachments.get[ClassForCaseCompanionAttachment] match {
+                case Some(att) =>
+                  val cdef = att.caseClass
+                  stats.exists {
```

Comment:
Thanks for explaining, I start to understand a bit more what's happening. If I understand correctly, this PR tightens the conditions under which such synthetics are generated, presumably making it "deterministic" -- the same place where the ClassDef is. But where were they added before this PR? Maybe there's an equivalent condition that doesn't require iterating?

I agree the code won't run for all stats in a package, but the typer is generally "hot", and even if this doesn't seem much, it's still going to be doing more work than before, work that's not a constant but a function of code size.